### PR TITLE
Get user snapshots dataframe

### PR DESCRIPTION
**Pull request by andrew8bit**
- [Upgrade indico-client 4.9.0 > 5.0.1](https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/commit/d3f36c9f0532f112d91eed5812a521350c3ff3e2)[](https://github.com/andrew8bit)
 
- [Created Usermetrics class](https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/commit/a3586759313930f4f020be3b58ee4425828c2eb9)[](https://github.com/andrew8bit) 

- [Function to write user metrics to disk](https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/commit/ed68eb4142f982ea50ec1c6a9a4ad0bd144f0557)[](https://github.com/andrew8bit)

- [Importing UserMetrics](https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/commit/26bfb68e09242662b87d55cef764917847e1c351)[](https://github.com/andrew8bit)

- [Test - Added tests for user metrics](https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/commit/e47d1ed08cfa2f6036507705f061cb08394e7f4c)[](https://github.com/andrew8bit)
 
- [Example - Added an example of storing user metrics df and writing to …](https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/commit/9558d20ea14d572958501eb2eaee919d11b81b93)[](https://github.com/andrew8bit)

### CODE REVIEW

**Comment by Scott771** [comment]
Hey- checkout the version of "black" that is in main, build is failing because of:

`Run black indico_toolkit
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.[12](https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/runs/6017779550?check_suite_focus=true#step:5:12)/x64/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/black/__init__.py", line 6985, in patched_main
    patch_click()
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/black/__init__.py", line 6974, in patch_click
    from click import _unicodefun  # type: ignore
ImportError: cannot import name '_unicodefun' from 'click' (/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/click/__init__.py)`

File: indico_toolkit/metrics/usermetrics.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+from typing import List, Dict
+from indico import IndicoClient
+from datetime import datetime
+from indico_toolkit.indico_wrapper import IndicoWrapper
+from indico_toolkit import ToolkitInputError
+from indico.queries.usermetrics import GetUserSnapshots
+from indico.filters import UserMetricsFilter
+
+
+class UserMetrics(IndicoWrapper):
```
Comment (nateshim-indico) [inline indico_toolkit/metrics/usermetrics.py:11]: maybe we can add 1 sentence docstring summarizing what this class does like in FileProcessing or AutoReviewer

File: indico_toolkit/metrics/usermetrics.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            )
+
+        all_user_metrics = []
+        for snapshots in self.client.paginate(
+            GetUserSnapshots(date=date, filters=filter_opts)
+        ):
+            for snapshot in snapshots:
+                user_metrics = []
+                for column in columns:
+                    if column == "datasets":
```
Comment (nateshim-indico) [inline indico_toolkit/metrics/usermetrics.py:51]: I think we can shorten this to if "datasets" in columns

File: indico_toolkit/metrics/usermetrics.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+from indico.filters import UserMetricsFilter
+
+
+class UserMetrics(IndicoWrapper):
+    def __init__(self, client: IndicoClient):
+        self.client = client
+
+    def get_user_metrics_df(
+        self,
+        columns: List[str] = [
```
Comment (nateshim-indico) [inline indico_toolkit/metrics/usermetrics.py:17]: could/should we make these arguments instance variables? It's just that I'm seeing columns, date, filter_user_id, and filter_email being used in both functions.

File: tests/metrics/test_usermetrics.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+from typing import List
+
+from indico_toolkit.metrics import UserMetrics
+
+
+@pytest.fixture(scope="module")
+def ex_user_metrics_object(indico_client):
+    user_metrics = UserMetrics(indico_client)
+    return user_metrics
+
```
Comment (nateshim-indico) [inline tests/metrics/test_usermetrics.py:13]: I think we should add some tests for get_user_metrics_df as well

File: examples/user_metrics.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+HOST = "app.indico.io"
+API_TOKEN_PATH = "./indico_api_token.txt"
+
+client = create_client(HOST, API_TOKEN_PATH)
+
+"""
+Example 1: Create a dataframe of user metrics with columns id, name, and email
+"""
+user_metrics = UserMetrics(client)
+df = user_metrics.get_user_metrics_df(columns=["id", "name", "email"])
```
Comment (fitzworkhub) [inline examples/user_metrics.py:17]: How will users know what columns they can pull from?

File: indico_toolkit/metrics/usermetrics.py (hunk 1 above)
Comment (fitzworkhub) [reply indico_toolkit/metrics/usermetrics.py:11]: I agree with this

File: indico_toolkit/metrics/usermetrics.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        filter_opts = {}
+        if filter_user_id or filter_email:
+            filter_opts = UserMetricsFilter(
+                user_id=filter_user_id, filter_email=filter_email
+            )
+
+        all_user_metrics = []
+        for snapshots in self.client.paginate(
+            GetUserSnapshots(date=date, filters=filter_opts)
+        ):
```
Comment (fitzworkhub) [inline indico_toolkit/metrics/usermetrics.py:47]: This should be refactored to another member function `get_user_snapshots`

File: indico_toolkit/metrics/usermetrics.py (hunk 3 above)
Comment (fitzworkhub) [reply indico_toolkit/metrics/usermetrics.py:17]: I agree with Nate. This should be defined as a list at the top of the module.

File: indico_toolkit/metrics/usermetrics.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+from indico_toolkit import ToolkitInputError
+from indico.queries.usermetrics import GetUserSnapshots
+from indico.filters import UserMetricsFilter
+
+
+class UserMetrics(IndicoWrapper):
+    def __init__(self, client: IndicoClient):
+        self.client = client
+
+    def get_user_metrics_df(
```
Comment (fitzworkhub) [inline indico_toolkit/metrics/usermetrics.py:15]: See notes about refactoring below. This function might be better a property functiony like: `user_metrics.dataframe`

File: indico_toolkit/metrics/usermetrics.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        """
+        Write a CSV to disc of user metrics from the cluster
+        Args:
+            output_path (str): path to write CSV on your system, e.g. "./user_metrics.csv"
+            columns (List[str]): List of columns to include in the dataframe (OPTIONAL) - defaults to all
+            date (datetime): specific date to query - defaults to current datetime (OPTIONAL)
+            filter_user_id (int): user_id to filter for (OPTIONAL)
+            filter_email (str): user email to filter for (OPTIONAL)
+        """
+        df = self.get_user_metrics_df(columns, date, filter_user_id, filter_email)
```
Comment (fitzworkhub) [inline indico_toolkit/metrics/usermetrics.py:91]: after refactoring you don't want to call this in here. It should only save out the dataframe attribute.

File: tests/metrics/test_usermetrics.py (hunk 4 above)
Comment (fitzworkhub) [reply tests/metrics/test_usermetrics.py:13]: We definitely need a test post refactor. You might want to mock some sample I/O. 

We need some kind of example of what the output data looks like too.

**Review by fitzworkhub — CHANGES_REQUESTED** [review]
Added a couple of refactoring and testing requests.

**Comment by Scott771** [comment]
@andrew8bit  if you resolve something, you should make a note of the change you made and click "resolve"

File: examples/user_metrics.py (hunk 5 above)
Comment (andrew8bit) [reply examples/user_metrics.py]: Resolved - removed options for columns - users get all the information back now

File: indico_toolkit/metrics/usermetrics.py (hunk 1 above)
Comment (andrew8bit) [reply indico_toolkit/metrics/usermetrics.py:25]: Updated here - https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/blob/e49e5e55a1c12d9a696330b5f1355051839ed5b2/indico_toolkit/metrics/usermetrics.py#L26

File: indico_toolkit/metrics/usermetrics.py (hunk 3 above)
Comment (andrew8bit) [reply indico_toolkit/metrics/usermetrics.py]: Removed option for columns in methods - static list on top of module now - doesn't save us any resources by decreasing any return data, so we're just returning everything now

File: tests/metrics/test_usermetrics.py (hunk 4 above)
Comment (andrew8bit) [reply tests/metrics/test_usermetrics.py:24]: Resolved here - https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/blob/e49e5e55a1c12d9a696330b5f1355051839ed5b2/tests/metrics/test_usermetrics.py#L26

File: indico_toolkit/metrics/usermetrics.py (hunk 6 above)
Comment (andrew8bit) [reply indico_toolkit/metrics/usermetrics.py]: Resolved here - https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/blob/e49e5e55a1c12d9a696330b5f1355051839ed5b2/indico_toolkit/metrics/usermetrics.py#L50

File: indico_toolkit/metrics/usermetrics.py (hunk 7 above)
Comment (andrew8bit) [reply indico_toolkit/metrics/usermetrics.py]: Added to class attributes here - resolved - https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/blob/e49e5e55a1c12d9a696330b5f1355051839ed5b2/indico_toolkit/metrics/usermetrics.py#L47

File: indico_toolkit/metrics/usermetrics.py (hunk 8 above)
Comment (andrew8bit) [reply indico_toolkit/metrics/usermetrics.py]: Saved to self.df here - resolved - https://github.com/IndicoDataSolutions/Indico-Solutions-Toolkit/blob/e49e5e55a1c12d9a696330b5f1355051839ed5b2/indico_toolkit/metrics/usermetrics.py#L116

File: indico_toolkit/metrics/usermetrics.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        """
+        filter_user_id(int): ID of a user to filter for
+        filter_email(str): Email of a user to filter for
+        If user_id and email are not passed, defaults to all available users
+        """
+        self.client = client
+        self.datetime = datetime
+        self.filter_user_id = filter_user_id
+        self.filter_email = filter_email
+        self.df = pd.DataFrame
```
Comment (Scott771) [inline indico_toolkit/metrics/usermetrics.py:47]: Should this be `self.df: pd.DataFrame = None`?

**Comment by Scott771** [comment]
@andrew8bit should we keep this? I'd test and see if it still works, otherwise can close
